tests/test_unittest/test_storage.py

```diff
@@ -81,99 +71,108 @@ def allocation_strategy(draw):
 
 
 @hyp_st.composite
-def mask_strategy(draw):
-    dimension = draw(hyp_st.integers(min_value=0, max_value=4))
+def axes_strategy(draw):
+    dimension = draw(hyp_st.integers(min_value=0, max_value=3))
 
-    mask_strats = []
-    shape_strats = []
-    default_origin_strats = []
+    shape = draw(
+        hyp_st.tuples(*[hyp_st.integers(min_value=0, max_value=10) for i in range(dimension)])
+    )
+    default_origin = draw(
+        hyp_st.tuples(
+            *[
+                hyp_st.integers(min_value=0, max_value=max(0, min(5, shape[i] - 1)))
+                for i in range(dimension)
+            ]
+        )
+    )
 
-    for i in range(dimension):
-        shape_strats = shape_strats + [hyp_st.integers(min_value=1, max_value=64)]
-        mask_strats = mask_strats + [hyp_st.booleans()]
-    shape = draw(hyp_st.tuples(*shape_strats))
-    for i in range(dimension):
-        default_origin_strats = default_origin_strats + [
-            hyp_st.integers(min_value=0, max_value=min(32, shape[i] - 1))
-        ]
-    default_origin = draw(hyp_st.tuples(*default_origin_strats))
-    mask = draw(
+    axes = draw(
         hyp_st.one_of(
             hyp_st.just(None),
-            hyp_st.tuples(*mask_strats),
-            hyp_st.permutations(
-                ([True] * dimension) + ([False] * draw(hyp_st.integers(min_value=0, max_value=10)))
+            hyp_st.lists(hyp_st.booleans(), min_size=3, max_size=3).filter(
+                lambda x: sum(x) == dimension
             ),
         )
     )
-    return dict(shape=shape, default_origin=default_origin, mask=mask)
+    if isinstance(axes, list):
+        axes = [a for m, a in zip(axes, "IJK") if m]
+    return dict(shape=shape, default_origin=default_origin, axes=axes)
 
 
 # ---- Tests ----
+def test_allocate_cpu():
+    run_test_allocate_singledevice(allocate_function=allocate_cpu)
+
+
+@pytest.mark.requires_gpu
+def test_allocate_gpu_only():
+    run_test_allocate_singledevice(allocate_function=allocate_gpu_only)
+
+
 @hyp.given(param_dict=allocation_strategy())
-def test_allocate_cpu(param_dict):
+def run_test_allocate_singledevice(param_dict, allocate_function):
     alignment_bytes = param_dict["alignment"]
     dtype = param_dict["dtype"]
     default_origin = param_dict["default_origin"]
     shape = param_dict["shape"]
     layout_map = param_dict["layout_order"]
+    ndim = len(shape)
 
-    raw_buffer, field = allocate_cpu(default_origin, shape, layout_map, dtype, alignment_bytes)
-
-    # check that field is a view of raw_buffer
-    assert field.base is raw_buffer
+    raw_buffer, field = allocate_function(
+        default_origin, shape, layout_map, dtype, alignment_bytes
```

Comment:
It looks like none of the tests set `data`. This should probably be tested as it is what differentiates e.g. `zeros` from `ones`.

Reply:
These tests also don't seem to test `default_parameters`, `axes`, `copy`, or `device_data`.